### PR TITLE
DSCS-2477 adds a light-weight summary entity for Works; adds a method…

### PR DESCRIPTION
**Pull request by m-r-c**
… to retrieve the hierarchy leading to a work (used for the breadcrumbs in banjo)

@scoen @jrixon @sjacob @ninhnguyen @shuangzhou @weijunnla @adityaburra

### CODE REVIEW

**Comment by nlabuilder** [comment]
SonarQube analysis reported 5 issues

1. MAJOR [AmberSession.java](https://github.com/nla/amberdb/blob/b379b8596b94d0d10d8e37ef0edd2270a839c924/src/main/java/amberdb/AmberSession.java): This file has 1,074 lines, which is greater than 1,000 authorized. Split it into smaller files. [rule](https://sonarqube.nla.gov.au/coding_rules#rule_key=squid%3AS00104)
1. MAJOR [WorkSummary.java#L33](https://github.com/nla/amberdb/blob/b379b8596b94d0d10d8e37ef0edd2270a839c924/src/main/java/amberdb/model/WorkSummary.java#L33): Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarqube.nla.gov.au/coding_rules#rule_key=squid%3AS1186)
1. MINOR [AmberSession.java#L301](https://github.com/nla/amberdb/blob/b379b8596b94d0d10d8e37ef0edd2270a839c924/src/main/java/amberdb/AmberSession.java#L301): Define a constant instead of duplicating this literal "collection" 3 times. [rule](https://sonarqube.nla.gov.au/coding_rules#rule_key=squid%3AS1192)
1. MINOR [AmberSession.java#L335](https://github.com/nla/amberdb/blob/b379b8596b94d0d10d8e37ef0edd2270a839c924/src/main/java/amberdb/AmberSession.java#L335): Define a constant instead of duplicating this literal "bibId" 3 times. [rule](https://sonarqube.nla.gov.au/coding_rules#rule_key=squid%3AS1192)
1. MINOR [WorkSummary.java#L10](https://github.com/nla/amberdb/blob/b379b8596b94d0d10d8e37ef0edd2270a839c924/src/main/java/amberdb/model/WorkSummary.java#L10): Document this public class. [rule](https://sonarqube.nla.gov.au/coding_rules#rule_key=squid%3AUndocumentedApi)

**Comment by nlabuilder** [comment]
Build finished.

File: src/main/java/amberdb/AmberSession.java [hunk 1]
```diff
@@ -273,6 +257,96 @@ public Work findWork(String idOrAlias) {
...
+                    "           child.subUnitType,  \n" +
+                    "           child.subUnitNo,  \n" +
+                    "           child.sensitiveMaterial,  \n" +
+                    "           child.digitalStatus,  \n" +
+                    "           child.digitalStatusDate, \n" +
+                    "           children.childCount \n" +
+                    " from work child \n" +
+                    "   left join flatedge e on child.id = e.v_out and e.label = 'isPartOf' \n" +
+                    "   left join work parent on e.v_in = parent.id \n" +
+                    "   left join (select count(*) childCount, e2.v_in from flatedge e2 where e2.v_in = :id and e2.label = 'isPartOf' group by e2.v_in) children on children.v_in = child.id \n" +
```
Comment (shuangzhou) [inline src/main/java/amberdb/AmberSession.java]: The childCount might not need to include section work records such as articles and chapters.

Comment (shuangzhou) [inline src/main/java/amberdb/AmberSession.java]: What if the input work record is a top record without parent?

File: src/main/java/amberdb/AmberSession.java [hunk 2]
```diff
@@ -273,6 +257,96 @@ public Work findWork(String idOrAlias) {
...
+        summary.setDateCreated((Date) work.get("dateCreated"));
+        summary.setRecordSource((String) work.get("recordSource"));
+        summary.setLocalSystemNumber((String) work.get("localSystemNumber"));
+        summary.setBibLevel((String) work.get("bibLevel"));
+        summary.setBibId((String) work.get("bibId"));
+        summary.setSubUnitType((String) work.get("subUnitType"));
+        summary.setSubUnitNumber((String) work.get("subUnitNo"));
+        summary.setSensitiveMaterial((String) work.get("sensitiveMaterial"));
+        summary.setObjectId(PIUtil.format((Long) work.get("childId")));
+        summary.setChildCount(work.get("childCount") == null ? 0 : (Long) work.get("childCount"));
```
Comment (shuangzhou) [inline src/main/java/amberdb/AmberSession.java]: some of the above fields might not have a value e.g. creator, subUnitNo or sensitiveMaterial, should a value be set in the summary for these fields in this case?

**Comment by shuangzhou** [comment]
👍

File: src/main/java/amberdb/AmberSession.java (hunk 1 above)
Comment (m-r-c) [inline src/main/java/amberdb/AmberSession.java]: I have updated the query to exclude Sections from the child count.

The getHierarchySummaryForWork method will always return at least 1 item, which is the requested work (assuming it exists). If it has no parent then the parentId field will be null.

**Pull request merged**